src/systems/rf_comms/RFComms.cc

```diff
@@ -0,0 +1,457 @@
+}
+
+/////////////////////////////////////////////
+RFPower RFComms::Implementation::LogNormalReceivedPower(
+  const double &_txPower, const RadioState &_txState,
+  const RadioState &_rxState) const
+{
+  const double kRange = _txState.pose.Pos().Distance(_rxState.pose.Pos());
+
+  if (this->rangeConfig.maxRange > 0.0 && kRange > this->rangeConfig.maxRange)
+    return {-std::numeric_limits<double>::infinity(), 0.0};
+
+  const double kPL = this->rangeConfig.l0 +
+    10 * this->rangeConfig.fadingExponent * log10(kRange);
+
+  return {_txPower - kPL, this->rangeConfig.sigma};
+}
+
+/////////////////////////////////////////////
+std::tuple<bool, double> RFComms::Implementation::AttemptSend(
+  RadioState &_txState, RadioState &_rxState, const uint64_t &_numBytes)
+{
+  double now = _txState.timeStamp;
+
+  // Maintain running window of bytes sent over the last epoch, e.g., 1s.
+  while (!_txState.bytesSent.empty() &&
+         _txState.bytesSent.front().first < now - this->epochDuration)
+  {
+    _txState.bytesSentThisEpoch -= _txState.bytesSent.front().second;
+    _txState.bytesSent.pop_front();
+  }
+
+  // igndbg << "Bytes sent: " <<  _txState.bytesSentThisEpoch << " + "
+  //        << _numBytes << " = "
+  //        << _txState.bytesSentThisEpoch + _numBytes << std::endl;
+
+  // Compute prospective accumulated bits along with time window
+  // (including this packet).
+  double bitsSent = (_txState.bytesSentThisEpoch + _numBytes) * 8;
+
+  // Check current epoch bitrate vs capacity and fail to send accordingly
+  if (bitsSent > this->radioConfig.capacity * this->epochDuration)
+  {
+    ignwarn << "Bitrate limited: " << bitsSent << "bits sent (limit: "
+            << this->radioConfig.capacity * this->epochDuration << std::endl;
+    return std::make_tuple(false, std::numeric_limits<double>::lowest());
+  }
+
+  // Record these bytes.
+  _txState.bytesSent.push_back(std::make_pair(now, _numBytes));
+  _txState.bytesSentThisEpoch += _numBytes;
+
+  // Get the received power based on TX power and position of each node.
+  auto rxPowerDist =
+    this->LogNormalReceivedPower(this->radioConfig.txPower, _txState, _rxState);
+
+  double rxPower = rxPowerDist.mean;
+  if (rxPowerDist.variance > 0.0)
+  {
+    std::normal_distribution<> d{rxPowerDist.mean, sqrt(rxPowerDist.variance)};
+    rxPower = d(this->rndEngine);
+  }
+
+  // Based on rx_power, noise value, and modulation, compute the bit
+  // error rate (BER).
+  double ber = this->QPSKPowerToBER(
+    this->DbmToPow(rxPower), this->DbmToPow(this->radioConfig.noiseFloor));
+
+  double packetDropProb = 1.0 - exp(_numBytes * log(1 - ber));
+
+  // igndbg << "TX power (dBm): " << this->radioConfig.txPower << "\n" <<
+  //           "RX power (dBm): " << rxPower << "\n" <<
+  //           "BER: " << ber << "\n" <<
+  //           "# Bytes: " << _numBytes << "\n" <<
+  //           "PER: " << packetDropProb << std::endl;
+
+  double randDraw = (rand() % 1000) / 1000.0;
```

Comment:
one minor code check issue here:

```
/github/workspace/src/systems/rf_comms/RFComms.cc:292:  Consider using rand_r(...) instead of rand(...) for improved thread safety
```

Reply:
Replaced with an Ignition Math function in [e39c85d](https://github.com/ignitionrobotics/ign-gazebo/pull/1428/commits/e39c85ddf3dbf2aec8c60792b282602c9654b420).